go/vt/vtexplain/vtexplain.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2017 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package vtexplain analyzes a set of sql statements and returns the
+// corresponding vtgate and vttablet query plans that will be executed
+// on the given statements
+package vtexplain
+
+// XXX TODO:
+//
+// Add full grammar support for CREATE TABLE
+//   - reserved vs non-reserved keywords
+// For DML queries handle comments to indicate whether rows exist or not
+// Human-friendly and json output modes
+// Fix TabletQuery to pre-render BindVariables
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	log "github.com/golang/glog"
+
+	"github.com/youtube/vitess/go/vt/sqlparser"
+	"github.com/youtube/vitess/go/vt/vtgate/engine"
+
+	querypb "github.com/youtube/vitess/go/vt/proto/query"
+)
+
+// Options to control the explain process
+type Options struct {
+	// NumShards indicates the number of shards in the topology
+	NumShards int
+
+	// ReplicationMode must be set to either "ROW" or "STATEMENT" before
+	// initialization
+	ReplicationMode string
+
+	// Normalize controls whether or not vtgate does query normalization
+	Normalize bool
+}
+
+// TabletQuery defines a query that was sent to a given tablet and how it was
+// processed in mysql
+type TabletQuery struct {
+	// SQL command sent to the given tablet
+	SQL string
+
+	// BindVars sent with the command
+	BindVars map[string]*querypb.BindVariable
+
+	// The actual queries executed by mysql
+	MysqlQueries []string
+}
+
+// MarshalJSON renders the json structure
+func (tq *TabletQuery) MarshalJSON() ([]byte, error) {
+	// Convert Bindvars to strings for nicer output
+	bindVars := make(map[string]string)
+	for k, v := range tq.BindVars {
+		bindVars[k] = v.String()
```

Comment:
Good idea. I just used `EncodeValue` directly and added a test case to show this works for tuples.